torch/csrc/jit/codegen/cuda/scheduler/reduction_utils.cpp

```diff
@@ -0,0 +1,648 @@
+          auto pos = std::distance(rfactor_tv_dom.begin(), reduction_it);
+          // I would like computeAtMode here to be Standard. However, the
+          // processing of welford rfactors in compute at ends up propating
+          // compute at from reduction_tv->rfactor_tv to all outputs.
+          rfactor_tv->computeWith(
+              reduction_tvs[i], pos, ComputeAtMode::BestEffort);
+        } else {
+          rfactor_tvs[i]->computeWith(
+              reduction_tvs[i], -1, ComputeAtMode::BestEffort);
+        }
+      }
+    }
+
+    // Remove anything before a reduction from compute_from
+    {
+      auto producers_of_reductions = DependencyCheck::getAllValsBetween(
+          {fusion->inputs().begin(), fusion->inputs().end()},
+          {reduction_tvs.begin(), reduction_tvs.end()});
+
+      auto producer_tvs_of_reductions =
+          ir_utils::filterByType<TensorView>(producers_of_reductions);
+      compute_from.erase(
+          std::remove_if(
+              compute_from.begin(),
+              compute_from.end(),
+              [&producer_tvs_of_reductions](TensorView* compute_from_tv) {
+                return std::find(
+                           producer_tvs_of_reductions.begin(),
+                           producer_tvs_of_reductions.end(),
+                           compute_from_tv) != producer_tvs_of_reductions.end();
+              }),
+          compute_from.end());
+    }
+
+    // Add reduction tensor views to compute from
+    compute_from.insert(
+        compute_from.end(), reduction_tvs.begin(), reduction_tvs.end());
+
+    // Compute between reductions and output caches
+    scheduler_utils::computeAtBetween(
+        compute_from,
+        compute_to,
+        -1,
+        ComputeAtMode::BestEffort,
+        mapped_to_trivial_reduction);
+
+  } else {
+    // Want to inline, especially backwards based on reduction_tv, otherwise
+    // rfactor tv may not be inlined correctly
+    auto ref_tvs = rfactor_tvs.size() ? rfactor_tvs : reduction_tvs;
+    for (auto red_tv : ref_tvs) {
+      auto pos_it = std::find_if(
+          red_tv->domain()->domain().begin(),
+          red_tv->domain()->domain().end(),
+          [&mapped_to_trivial_reduction](IterDomain* id) {
+            return id->getParallelType() == ParallelType::Unswitch ||
+                id->getParallelType() == ParallelType::Unroll ||
+                id->getParallelType() == ParallelType::Vectorize ||
+                id->getParallelType() == ParallelType::MisalignedVectorize ||
+                mapped_to_trivial_reduction.count(id);
+          });
+      auto pos = pos_it == red_tv->domain()->domain().end()
+          ? -1
+          : std::distance(red_tv->domain()->domain().begin(), pos_it) + 1;
+
+      scheduler_utils::computeAtInputs(red_tv, pos, ComputeAtMode::MostInlined);
+      scheduler_utils::computeWithOutputs(
+          red_tv, pos, ComputeAtMode::BestEffort);
+    }
+  }
+}
+
+namespace {
+struct id_lt {
```

Comment:
Sorting function for id's before rfactor, coordinated with expectations from the inliner.